### ISSUE TITLE
is_bam is not accurate

### ISSUE THREAD

**Issue by vishnubob** [pysam-developers/pysam#186]
It would appear that the attribute `AlignmentFile.is_bam` is only reflecting wether the `b` flag used to open a given SAM file, instead of testing the content of the file and ascertaining if it is actually a BAM.  As an example:

```
>>> sam = pysam.AlignmentFile("example.sam", 'rb')
>>> sam.is_bam
True
>>> sam = pysam.AlignmentFile("example.sam", 'r')
>>> sam.is_bam
False
```

Since BAM files are built with a standard [BZGF header](https://en.wikipedia.org/wiki/Gzip#File_format), would it not make sense to test the file for the GZIP magic string `0x1F8B` at the start of the file?

**Comment by AndreasHeger** [comment]
Thanks, auto-detection is currently not implemented - something for the todo list.

Best wishes,
Andreas

**Comment by AndreasHeger** [comment]
Thanks, is_bam and is_cram are now set from htslib information.

**Comment by jmarshall** [comment]
Do `is_bam`, `is_cram` etc in chtslib.pxd's `ctypedef struct htsFile` actually work?  They are bit-fields in htslib all in either the first or last byte (who knows or cares :smile:), so I suspect the `ctypedef` ones won't be set correctly.

Whether they work or not, it would be better to use `hts_get_format()->format` (introduced in htslib 1.2) instead.  The `htsFile` struct is intended to be an opaque structure for use only internally within htslib, and I will clarify the comments in _hts.h_ to spell this out.  You should be able to have just an opaque `ctypedef struct htsFile` in _chtslib.pxd_, and that will insulate pysam from future htslib changes in this area.

**Comment by AndreasHeger** [comment]
Thanks, I added hts_get_format() as a property. I tried using htsFormat and htsExactFormat, but had issues with the cython mapping.

is_bam/is_cram do work (bitfields had been an issue, but newer cython is able to work around these, however I have not tested on many platforms).  The two flags can distinguish between SAM, BAM and CRAM and are covered by unit tests. I will move to using hts_get_format().
